sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/SynchronizedAccessor.java

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.identity.implementation;
+
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.ReplayProcessor;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+/**
+ * Synchronizes reactor threads accessing/instantiating a common value {@code T}.
+ *
+ * @param <T> The value being instantiated / accessed.
+ */
+public class SynchronizedAccessor<T> {
+    private final AtomicBoolean wip;
+    private T cache;
+    private final ReplayProcessor<T> emitterProcessor = ReplayProcessor.create(1);
+    private final FluxSink<T> sink = emitterProcessor.sink(FluxSink.OverflowStrategy.BUFFER);
+    private final Supplier<T> supplier;
+
+    public SynchronizedAccessor(Supplier<T> supplier) {
+        this.wip = new AtomicBoolean(false);
+        this.supplier = supplier;
+    }
+
+    /**
+     * Get the value from the configured supplier.
+     *
+     * @return the output {@code T}
+     */
+    public Mono<T> getValue() {
+        if (cache != null) {
+            return Mono.just(cache);
+        }
+        return Mono.defer(() -> {
+            if (!wip.getAndSet(true)) {
+                try {
+                    cache = supplier.get();
+                    sink.next(cache);
+                } catch (Exception e) {
+                    sink.error(e);
+                } finally {
+                    wip.set(false);
```

Comment:
There's still a chance of race condition here when a thread marks `wip` as false and then another thread enters the `if` block and calls `supplier.get()` again updating the reference to `cache`. How critical is the need to call the supplier only once?

Reply:
moved it to the defer block.

Reply:
If the expectation is to create a new `IdentityClient` in case of errors, there's no need to reset `wip` to false. 

Only one thread ever has to enter the `wip` block and should either get `value` from supplier or throw an error. There's no need for another thread to re-enter this block.